lib/updates.mli

```diff
@@ -0,0 +1,83 @@
+(** This module is a helper for interfaces of a particular shape.
+ *
+ * The idea is that updates are generated by a component when another
+ * component should be notified of events. For example, when a VM changes
+ * state, xenopsd uses this interface so that xapi knows something has
+ * changed. The updates contain no information other than that something
+ * has changed, so additionally the interface needs to have some sort of
+ * mechanism for querying the current state of the object in question.
+ *
+ * This module provides functions for adding updates, adding barriers
+ * (which contain a snapshot of the current updates), and a blocking call to
+ * that returns the latest set of updates, optionally given a token that
+ * represents the last set of updates received.
+*)
+
+module type INTERFACE =
+sig
+  val service_name : string
+  module Dynamic :
+  sig type id val rpc_of_id : id -> Rpc.t val id_of_rpc : Rpc.t -> id end
+end
+
+module Updates :
+  functor (Interface : INTERFACE) ->
+  sig
+    type id = int
+
+    (* Updates are kept in a `t` *)
+    type t
+
+    (* Create an empty updates recorder using the given scheduler *)
+    val empty : Scheduler.t -> t
+
+    (* last_id [dbg] [t] returns the last id that's in use in [t] *)
+    val last_id : string -> t -> int
+
+    (* The return type for `get`. Returns a list of barriers, then a list of updates, then a token
+       to pass to the next invocation of `get` *)
+    type get_result = (int * Interface.Dynamic.id list) list * Interface.Dynamic.id list * id
```

Comment:
Again, this is part of the interface, so difficult to change without causing backwards-incompatible breakage.